### PR TITLE
Ttag

### PR DESCRIPTION
**Pull request by pavloruban1**
add tag

### CODE REVIEW

File: server/src/app.ts [hunk 1]
```diff
@@ -6,6 +6,7 @@ import cors from 'cors';
 import { notFound } from './shared/not-found';
 import { errorHandler } from './error-handlers/error-handler';
 import Router from './router';
+import tagRouter from './tag/tag.router';
```
Comment (Simkav) [inline server/src/app.ts]: remove

File: server/src/app.ts [hunk 2]
```diff
@@ -14,6 +15,7 @@ app.use(helmet());
 app.use(cors());
 app.use(express.json());
 app.use('/api/v1', Router);
+app.use('/api', tagRouter);
```
Comment (Simkav) [inline server/src/app.ts]: remove

File: server/src/tag/tag.entity.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    is_specialization!: boolean;
+
+    @ManyToMany(() => User, (user) => user.tags)
+    users!: User[];
+    
+    @ManyToMany(() => User, user => user.specializations)
+    specializedUsers!: User[];
+
+    @ManyToMany(() => User, user => user.technologies)
+    technologies!: User[];
```
Comment (Simkav) [inline server/src/tag/tag.entity.ts]: remove

File: server/src/tag/tag.entity.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    id!: number;
+
+    @Column()
+    name!: string;
+
+    @Column()
+    color!: string;
+
+    @Column()
+    is_specialization!: boolean;
```
Comment (Simkav) [inline server/src/tag/tag.entity.ts]: isSpecialization

File: server/src/tag/tag.router.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import {
+  createTagSchema,
+  editTagSchema,
+  tagIdParamSchema,
+  addTagToUserSchema
+} from '../tag/tag.schemas';
+
+const tagRouter = Router();
+tagRouter.use(getUserJWT);
+
```
Comment (Simkav) [inline server/src/tag/tag.router.ts:13]: remove /tags from all routes, leave only in export

File: server/src/user/user.entity.ts [hunk 6]
```diff
@@ -10,4 +11,16 @@ export class User {
 
   @Column({ select: false })
   password!: string;
+
+  @ManyToMany(() => Tag, (tag) => tag.users)
+  @JoinTable()
+  tags!: Tag[];
```
Comment (Simkav) [inline server/src/user/user.entity.ts]: remove

File: server/src/tag/tag.service.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+export const editTag = async (id: number, newName: string) => {
+  const tag = await tagRepository.findOneBy({ id });
+  if (!tag) throw new Error('Tag not found');
+  tag.name = newName;
+  return await tagRepository.save(tag);
+};
+
+export const addTagToUser = async (userId: number, tagId: number) => {
+  const userRepository = AppDataSource.getRepository(User);
+  const user = await userRepository.findOne({ where: { id: userId }, relations: ['tags'] });
```
Comment (Simkav) [inline server/src/tag/tag.service.ts]: findUserByIdWithTags in user service?

File: server/src/tag/tag.controller.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  res.json({ message: 'Tag deleted successfully' });
+};
+
+const editTag = async (req: Request, res: Response) => {
+  const tag = await tagService.editTag(Number(req.params.id), req.body.name);
+  res.json(tag);
+};
+
+const addTagToUser = async (req: Request, res: Response) => {
+  const { userId, tagId } = req.body;
```
Comment (Simkav) [inline server/src/tag/tag.controller.ts]: userId from jwt token not from body

File: server/src/tag/tag.schemas.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import Joi from 'joi';
+
+export const createTagSchema = Joi.object({
```
Comment (Simkav) [inline server/src/tag/tag.schemas.ts]: isSpecialization?

File: server/src/tag/tag.service.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import { AppDataSource } from '../db/data-source';
+import { Tag } from '../tag/tag.entity';
+import { User } from '../user/user.entity';
+
+const tagRepository = AppDataSource.getRepository(Tag);
+
+export const createTag = async (name: string) => {
+  const tag = tagRepository.create({ name });
```
Comment (Simkav) [inline server/src/tag/tag.service.ts]: see how created user in user service

File: server/src/tag/tag.router.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  tagController.createTag
+);
+
+tagRouter.delete(
+  '/tags/:id',
+  validator({ params: tagIdParamSchema }),
+  tagController.deleteTag
+);
+
+tagRouter.put(
```
Comment (Simkav) [inline server/src/tag/tag.router.ts]: put=>patch

**Review by Simkav — CHANGES_REQUESTED** [review]

File: server/src/app.ts (hunk 1 above)
Comment (pavloruban1) [reply server/src/app.ts]: fix

File: server/src/tag/tag.service.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  const userRepository = AppDataSource.getRepository(User);
+  const user = await userRepository.findOne({
+    where: { id: userId },
+    relations: ['tags'],
+  });
+
+  if (!user) throw new Error('User not found');
+  
+  return user;
+};
```
Comment (Simkav) [inline server/src/tag/tag.service.ts]: https://github.com/baza-trainee/baza-crm-backend/pull/5/files#r1614791924

**Pull request merged**